Support insert into system.zookeeper

hanfei1991 opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/run_check.py
-->
### Changelog category (leave one):
- New Feature


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
This PR aims to resolve #22130 which allows inserting into `system.zookeeper`. To simplify, this PR is designed as:

1. only allow to write (name, path, value), three columns must exist in the same insert statement.
2. allow recursive create if a path and its prefix of path don't exist.
3. No transaction size limitation.
4. Not allow "/" identifier in "name" attribution.
5. Allow "/a/b/c///d/" in "path", it will resolve to "/a/b/c/d" automatically.

In a same insert statement, transaction are ganranteed by zookeeper, but when reading data from zookeeper, changes may happen in the same time. So this is a "Read Commited" isolation level.
This PR uses a trie to cache all the path change to avoid creating path repeatedly. 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

hanfei1991 committed:
Support insert into system.zookeeper

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=37596) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=37596) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=37596) it.</sub>

hanfei1991 committed:
update test

hanfei1991 committed:
add zookeeper tag

hanfei1991 committed:
fix style check

hanfei1991 commented:
@KochetovNicolai @tavplubix Any suggestions to the fast check fail? I ran this test in my local enviroument:
```
root@figo-laptop-dell:~/ClickHouse# PATH=$PATH:./build/programs/ tests/clickhouse-test --zookeeper 02311_system_zookeeper_insert
Using queries from '/root/ClickHouse/tests/queries' directory
Using clickhouse-client as client program (expecting split build)
Connecting to ClickHouse server... OK

Running 1 stateless tests (MainProcess).

02311_system_zookeeper_insert:                                          [ OK ]

1 tests passed. 0 tests skipped. 0.20 s elapsed (MainProcess).
Won't run stateful tests because test data wasn't loaded.
All tests have finished.
```

and succeeded.

tavplubix commented:
See [runlog.log](https://s3.amazonaws.com/clickhouse-test-reports/37596/d7b372b73ab087d6c00d7cac1667662921c51d8d/fast_test__actions_/runlog.log):
```
2022-05-27 16:12:45 02311_system_zookeeper_insert:                                          [ UNKNOWN ] 0.00 sec. - Test internal error: 
2022-05-27 16:12:45 UnicodeDecodeError
2022-05-27 16:12:45 'utf-8' codec can't decode byte 0x91 in position 7514: invalid start byte
2022-05-27 16:12:45   File "/ClickHouse/tests/clickhouse-test", line 1040, in run
2022-05-27 16:12:45     result = self.process_result_impl(proc, stdout, stderr, total_time)
2022-05-27 16:12:45 
2022-05-27 16:12:45   File "/ClickHouse/tests/clickhouse-test", line 828, in process_result_impl
2022-05-27 16:12:45     diff = Popen(
2022-05-27 16:12:45 
2022-05-27 16:12:45   File "/usr/lib/python3.8/subprocess.py", line 1015, in communicate
2022-05-27 16:12:45     stdout = self.stdout.read()
2022-05-27 16:12:45 
2022-05-27 16:12:45   File "/usr/lib/python3.8/codecs.py", line 322, in decode
2022-05-27 16:12:45     (result, consumed) = self._buffer_decode(data, self.errors, final)
```

Seems like some of SELECT queries from the test returned binary data, so `clickhouse-test` cannot parse output of the test.

Try to run locally other tests that use zookeeper before running your test:
```
clickhouse-test --zookeeper -j 4 --order=random zoo
```

hanfei1991 commented:
Got it! The problem is the test outputs all the zookeeper content but they are not cleaned by previous tests. I will change the test.

hanfei1991 committed:
update tests

hanfei1991 committed:
fix flaky tests

hanfei1991 commented:
@tavplubix @KochetovNicolai  Hi, I failed the flaky test, could you please take a look of it?
in the flaky test, the diff of the result is 
```
2022-05-27 15:52:11 -/2-insert-testx	testz	z
2022-05-27 15:52:11 +/2-insert-testx	testz	y
```

However, the test is written as
```
insert into system.zookeeper (name, path, value) values ('testz', '/2-insert-testx', 'y');
insert into system.zookeeper (name, path, value) values ('testc', '2-insert-testz//c/cd/dd//', 'y');
insert into system.zookeeper (name, value, path) values ('testz', 'z', '/2-insert-testx');
```

 insert z happens after insert y. I am not familiar with ClickHouse insert mechanism. Are multiple inserts executing concurrently? Do you have any suggestions to look into it? Thanks a lot!

tavplubix commented:
It's because we run tests in parallel by default. You should add `no-parallel` tag or (much better) use unique path prefix for each test run. You can use current database name as unique value like this:
```
insert into system.zookeeper (name, path, value) values ('c', '/1-insert-testc/' || currentDatabase() ||  '/c/c/c/c/c/c', 11) ...
```

hanfei1991 commented:
> It's because we run tests in parallel by default. You should add `no-parallel` tag or (much better) use unique path prefix for each test run. You can use current database name as unique value like this:
> 
> ```
> insert into system.zookeeper (name, path, value) values ('c', '/1-insert-testc/' || currentDatabase() ||  '/c/c/c/c/c/c', 11) ...
> ```

We cannot check the unique path in the reference file though. I guess `no-parallel` is a cleaner way.

tavplubix commented:
> We cannot check the unique path in the reference file though

Actually we can, `clickhouse-test` replaces randomly generated database name (like `test_z20leh`) with `default` in test output before comparing it with reference file

hanfei1991 commented:
Perfect!

On Sat, May 28, 2022, 2:38 AM Alexander Tokmakov ***@***.***>
wrote:

> We cannot check the unique path in the reference file though
>
> Actually we can, clickhouse-test replaces randomly generated database
> name (like test_z20leh) with default in test output before comparing it
> with reference file
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ClickHouse/ClickHouse/pull/37596#issuecomment-1139930978>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE7KYQT6LTOZPP6FQ2DE4GLVMEJD5ANCNFSM5XD4MIEQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

hanfei1991 committed:
try fix flaky tests and refine code style

hanfei1991 committed:
fix build

hanfei1991 committed:
fix style

hanfei1991 commented:
@tavplubix In fact, `insert into values ("/" || currentDatabase() || "abc")` always insert "/abc", meaning currentDatabase() always return emtry string in insert statement. Is it by design?
To work around it, I insert the values into a temp table and select them by `select "/" || currentDatabase() || path` and then insert into zookeeper. It seems working well by far.

hanfei1991 commented:
@tavplubix the flaky tests fail again. The error message is:
```
2022-05-28 20:50:42 --- /usr/share/clickhouse-test/queries/0_stateless/02311_system_zookeeper_insert.reference	2022-05-28 20:48:11.361901234 +0930
2022-05-28 20:50:42 +++ /tmp/clickhouse-test/0_stateless/02311_system_zookeeper_insert.658.stdout	2022-05-28 20:50:41.884122509 +0930
2022-05-28 20:50:42 @@ -18,7 +18,7 @@
2022-05-28 20:50:42  /default/1-insert-testc/c/c/kk	g	14
2022-05-28 20:50:42  -------------------------
2022-05-28 20:50:42  /default/2-insert-testx	testc	x
2022-05-28 20:50:42 -/default/2-insert-testx	testz	z
2022-05-28 20:50:42 +/default/2-insert-testx	testz	y
2022-05-28 20:50:42  /default/2-insert-testz	c	
2022-05-28 20:50:42  /default/2-insert-testz/c	cd	
2022-05-28 20:50:42  /default/2-insert-testz/c/cd	dd	
2022-05-28 20:50:42 
2022-05-28 20:50:42 
2022-05-28 20:50:42 Settings used in the test: --max_insert_threads=14 --group_by_two_level_threshold=100000 --group_by_two_level_threshold_bytes=1 --distributed_aggregation_memory_efficient=1 --fsync_metadata=1 --output_format_parallel_formatting=0 --input_format_parallel_parsing=1 --min_chunk_bytes_for_parallel_parsing=10630691 --max_read_buffer_size=964684 --prefer_localhost_replica=1 --max_block_size=13784 --max_threads=54 --optimize_or_like_chain=0
2022-05-28 20:50:42 
2022-05-28 20:50:42 Database: test_crzwrw
2022-05-28 20:50:42 
2022-05-28 20:50:42 Running about 12 stateless tests (ForkPoolWorker-9).
2022-05-28 20:50:42 
```

Are insert statements running concurrently in the same sql file?

hanfei1991 committed:
try to make test stable

hanfei1991 commented:
@tavplubix anyway I refined the test and the flaky tests are still failing. Please see https://s3.amazonaws.com/clickhouse-test-reports/37596/d883194eb21ea53e32f7465e7064cf0d85edf49d/stateless_tests_flaky_check__address__actions_.html

However, I cannot figure out any info from the report. Could you please take a look?

tavplubix commented:
> `insert into values ("/" || currentDatabase() || "abc") always insert "/abc", meaning currentDatabase()` always return emtry string in insert statement. Is it by design?

No, it's a bug. 

> However, I cannot figure out any info from the report. Could you please take a look?

Open [runlog.log](https://s3.amazonaws.com/clickhouse-test-reports/37596/d883194eb21ea53e32f7465e7064cf0d85edf49d/stateless_tests_flaky_check__address__actions_/runlog.log) and and it will become clear:
```
2022-05-28 17:04:29 02311_system_zookeeper_insert:                                          [ FAIL ] 65.09 sec. - Test runs too long (> 60s). Make it faster.
2022-05-28 17:04:29 Settings used in the test: --max_insert_threads=0 --group_by_two_level_threshold=100000 --group_by_two_level_threshold_bytes=50000000 --distributed_aggregation_memory_efficient=1 --fsync_metadata=1 --output_format_parallel_formatting=0 --input_format_parallel_parsing=1 --min_chunk_bytes_for_parallel_parsing=13082811 --max_read_buffer_size=501217 --prefer_localhost_replica=1 --max_block_size=57938 --max_threads=12 --optimize_or_like_chain=0
2022-05-28 17:04:29 
2022-05-28 17:04:29 Database: test_3hiaaa
```

But probably it's because of `ThreadFuzzer` and parallel run, so you can simply ignore this failure.

tavplubix commented on src/Storages/System/StorageSystemZooKeeper.cpp:
```diff
@@ -28,6 +31,128 @@ namespace ErrorCodes
     extern const int BAD_ARGUMENTS;
 }
 
+// ZkNodeCache is a trie tree to cache all the zookeeper writes. The purpose of this struct is to avoid creating/setting nodes
```
Actually it's not "real" trie, because "real" trie should avoid creating intermediate nodes like this:
```
     A
    /  \
   B    X
  /
 C
 /
D
```

and should look like 
```
     A
   /   \
BCD      X
```
But "real" trie would be an overkill, so it's fine :)
And sorry for being bore :)

antonio2368 replied:
Isn't this actually a radix tree (https://en.wikipedia.org/wiki/Radix_tree), a special case of trie? 😄

tavplubix commented on src/Storages/System/StorageSystemZooKeeper.cpp:
```diff
@@ -28,6 +31,128 @@ namespace ErrorCodes
+
+    void insert(const std::vector<String> & nodes, zkutil::ZooKeeperPtr zookeeper, const String & value_to_set, size_t index)
+    {
+        if (index >= nodes.size())
```
Can `index` be greater than `nodes.size()`?

hanfei1991 replied:
No. So "==" is also fine, i'll fix it

tavplubix commented on src/Storages/System/StorageSystemZooKeeper.cpp:
```diff
@@ -28,6 +31,128 @@ namespace ErrorCodes
+        if (!children.contains(child_name))
+        {
+            String sub_path = "/" + boost::algorithm::join(std::vector<String>(nodes.begin(), nodes.begin() + index), "/");
+            bool ch_exist = false;
```
What does `ch` mean here? I suppose it means "child", but it may look like "ClickHouse" or something else :)

hanfei1991 replied:
changed to child_exist

tavplubix commented on src/Storages/System/StorageSystemZooKeeper.cpp:
```diff
@@ -28,6 +31,128 @@ namespace ErrorCodes
+
+    void generateRequests(Coordination::Requests & requests)
+    {
+        // If the node doesn't exists, we should generate create request.
```
Our codestyle recommends to start comments with 3 slashes: `///`
https://clickhouse.com/docs/en/development/style#comments
Actually there's no reason in this recommendation, but let's follow it for consistency.

hanfei1991 replied:
done

tavplubix commented on src/Storages/System/StorageSystemZooKeeper.cpp:
```diff
@@ -28,6 +31,128 @@ namespace ErrorCodes
+        // This dfs will prove ancestor nodes are processed first.
+        if (!exists)
+        {
+            auto request = std::make_shared<Coordination::CreateRequest>();
+            request->path = path;
+            request->data = value;
+            request->is_ephemeral = false;
+            request->is_sequential = false;
```
Use `zkutil::makeCreateRequest`

tavplubix commented on src/Storages/System/StorageSystemZooKeeper.cpp:
```diff
@@ -28,6 +31,128 @@ namespace ErrorCodes
+        }
+        else if (changed)
+        {
+            auto request = std::make_shared<Coordination::SetRequest>();
+            request->path = path;
+            request->data = value;
+            request->version = -1;
```
Use `zkutil::makeSetRequest`

tavplubix commented on src/Storages/System/StorageSystemZooKeeper.cpp:
```diff
@@ -28,6 +31,128 @@ namespace ErrorCodes
+            requests.push_back(request);
+        }
+        for (auto [_, child] : children)
+            child->generateRequests(requests);
```
Looks like recursion depth is not limited here and depends on user's input only. Should we call `checkStackSize()` here?

hanfei1991 replied:
Good point. I think the better way is to restrict the input path length, which impacts the stack size. What value do you think is ok?

tavplubix replied:
Maybe `PATH_MAX` (or just 4096)? But we need a test that tries to insert a sting like this

hanfei1991 replied:
Good idea

tavplubix commented:
Our RBAC works a bit weird with `system` database. For example, a user sill can read from system tables like `system.one` and `system.numbers` even if all grants are revoked. Please add a test that user without necessary grants cannot insert into `system.zookeeper` (just to be sure).

hanfei1991 committed:
Update src/Storages/System/StorageSystemZooKeeper.cpp

Co-authored-by: Alexander Tokmakov <tavplubix@gmail.com>

hanfei1991 committed:
Update src/Storages/System/StorageSystemZooKeeper.cpp

Co-authored-by: Alexander Tokmakov <tavplubix@gmail.com>

hanfei1991 committed:
address comments

hanfei1991 committed:
Merge branch 'hanfei/zk-write' of github.com:hanfei1991/ClickHouse into hanfei/zk-write

nvartolomei commented:
@tavplubix maybe this feature should be behind a feature flag? e.g. I would never want non-readonly XML users to write to zookeeper.

hanfei1991 commented:
@tavplubix could you provide some suggestions to write "user" case. I got stuck on it. For example, in default config, when I run some user case, it reports error:
```
root@figo-laptop-dell:~/ClickHouse# PATH=$PATH:~/ClickHouse/build/programs/ tests/clickhouse-test 01939_user_with_default_database
Using queries from '/root/ClickHouse/tests/queries' directory
Using clickhouse-client as client program (expecting split build)
Connecting to ClickHouse server... OK

Running 1 stateless tests (MainProcess).

01939_user_with_default_database:                                       [ FAIL ] - having stderror:
Received exception from server (version 22.6.1):
Code: 497. DB::Exception: Received from localhost:9000. DB::Exception: default: Not enough privileges. To execute this query it's necessary to have grant DROP USER ON *.*. (ACCESS_DENIED)
(query: drop user if exists u_01939)
Received exception from server (version 22.6.1):
Code: 497. DB::Exception: Received from localhost:9000. DB::Exception: default: Not enough privileges. To execute this query it's necessary to have grant CREATE USER ON *.*. (ACCESS_DENIED)
(query: create user u_01939 default database db_01939)
Code: 516. DB::Exception: Received from localhost:9000. DB::Exception: u_01939: Authentication failed: password is incorrect or there is no user with such name. (AUTHENTICATION_FAILED)

Received exception from server (version 22.6.1):
Code: 497. DB::Exception: Received from localhost:9000. DB::Exception: default: Not enough privileges. To execute this query it's necessary to have grant ALTER USER ON *.*. (ACCESS_DENIED)
(query: alter user u_01939 default database NONE)
Received exception from server (version 22.6.1):
Code: 497. DB::Exception: Received from localhost:9000. DB::Exception: default: Not enough privileges. To execute this query it's necessary to have grant SHOW USERS ON *.*. (ACCESS_DENIED)
(query: show create user u_01939)
Received exception from server (version 22.6.1):
Code: 497. DB::Exception: Received from localhost:9000. DB::Exception: default: Not enough privileges. To execute this query it's necessary to have grant ALTER USER ON *.*. (ACCESS_DENIED)
(query: alter user u_01939 default database `NONE`)
Received exception from server (version 22.6.1):
Code: 497. DB::Exception: Received from localhost:9000. DB::Exception: default: Not enough privileges. To execute this query it's necessary to have grant SHOW USERS ON *.*. (ACCESS_DENIED)
(query: show create user u_01939)
Received exception from server (version 22.6.1):
Code: 497. DB::Exception: Received from localhost:9000. DB::Exception: default: Not enough privileges. To execute this query it's necessary to have grant DROP USER ON *.*. (ACCESS_DENIED)
(query: drop user u_01939 )
```

What the config should I do in advance when I test these cases locally

hanfei1991 committed:
address comments

hanfei1991 commented on src/Storages/System/StorageSystemZooKeeper.cpp:
```diff
@@ -28,6 +31,128 @@ namespace ErrorCodes
+                if (path_vec[j].empty())
+                    path_vec.erase(path_vec.begin() + j);
+            }
+            path_vec.push_back(name);
```
I work around it by skipping all empty nodes when walking through trie

hanfei1991 commented on src/Storages/System/StorageSystemZooKeeper.cpp:
```diff
@@ -28,6 +31,128 @@ namespace ErrorCodes
+            for (int j = int(path_vec.size()) - 1; j >= 0; j--)
+            {
+                if (path_vec[j].empty())
+                    path_vec.erase(path_vec.begin() + j);
```
I work around it by skipping all empty nodes when walking through trie

hanfei1991 commented on src/Storages/System/StorageSystemZooKeeper.cpp:
```diff
@@ -28,6 +31,128 @@ namespace ErrorCodes
+// ZkNodeCache is a trie tree to cache all the zookeeper writes. The purpose of this struct is to avoid creating/setting nodes
+// repeatedly. For example, If we create path /a/b/c/d/e and path /a/b/d/f in the same transaction. We don't want to create
+// their common path "/a/b" twice. This data structure will cache this changes and generates the eventual requests within one pass.
+struct ZkNodeCache
```
:)

hanfei1991 commented on tests/queries/0_stateless/02311_system_zookeeper_insert.sql:
```diff
@@ -0,0 +1,37 @@
+insert into system.zookeeper (name, path, value) values ('a/b/c', '/', 'y'); -- { serverError 36 }
+insert into system.zookeeper (name, path, value) values ('/', '/a/b/c', 'z'); -- { serverError 36 }
+
+set allow_unrestricted_reads_from_keeper = 'false';
```
done

tavplubix commented:
Do you have `<access_management>1</access_management>` in `default` user's definition in your `users.xml`?

tavplubix commented:
> @tavplubix maybe this feature should be behind a feature flag? e.g. I would never want non-readonly XML users to write to zookeeper.

Makes sense, we should disable this feature by default, because it's too dangerous. For example, if someone can insert into `system.zookeeper` they can insert `GRANT` query into the queue of `ON CLUSTER` queries or change something in `ReplicatedAccessStorage` and get arbitrary grants. So looks like "user is allowed to insert into system.zookeeper" == "user is allowed to do anything". Maybe we can check access for inserting into `system.zookeeper` special way and allow it only if user has all privileges. @nvartolomei, what do you think?

hanfei1991 committed:
add user access tests

hanfei1991 commented:
> Our RBAC works a bit weird with `system` database. For example, a user sill can read from system tables like `system.one` and `system.numbers` even if all grants are revoked. Please add a test that user without necessary grants cannot insert into `system.zookeeper` (just to be sure).

Added, PTAL

tavplubix reviewed (approved):
LGTM, but let's add a configuration parameter that enables this feature (disabled by default). See `Context::getConfigRef()`

tavplubix commented:
Also `ClickHouse special build check (actions)` and `Style Check (actions)` failures are related (other failures are not)

hanfei1991 committed:
add config and fix style check

hanfei1991 commented:
> LGTM, but let's add a configuration parameter that enables this feature (disabled by default). See `Context::getConfigRef()`

Good Suggestion. I have added the config check and fixed the code style.

tavplubix commented on src/Core/Settings.h:
```diff
@@ -584,6 +584,8 @@ static constexpr UInt64 operator""_GiB(unsigned long long value)
     \
     M(Bool, allow_unrestricted_reads_from_keeper, false, "Allow unrestricted (without condition on path) reads from system.zookeeper table, can be handy, but is not safe for zookeeper", 0) \
     \
+    M(Bool, allow_writes_to_zookeeper, false, "Allow writes to system.zookeeper table, can be handy, but is not safe for zookeeper", 0) \
```
This does not protect from anything. Please, read previous comments in the conversation carefully.

hanfei1991 replied:
you mean a config item like "allow_zookeeper_write : true/false" in the config file? Well, I was too hurry

hanfei1991 replied:
Then how do we enable it in the tests?

hanfei1991 replied:
I found `01594_too_low_memory_limits.sh` also uses ad-hoc config file, I will look into it

tavplubix replied:
`01594_too_low_memory_limits` is bad test that was added by mistake, it should have been integration test. The correct way to enable it in tests is to add xml file to `tests/config/config.d/` and update `tests/config/install.sh`

hanfei1991 committed:
add config and change test logic

hanfei1991 commented:
@tavplubix I added the config `allow_zookeeper_write` for safeguard and adapted the test according to `01594_too_low_memory_limits`

hanfei1991 committed:
refine tests

hanfei1991 committed:
add zk_write.xml

tavplubix commented:
Hm, I've downloaded build from this PR, tried it locally and found some unexpected behavior:
```
dell9510 :) insert into system.zookeeper (name, version, czxid) values ('x', 1, 42)

INSERT INTO system.zookeeper (name, version, czxid) FORMAT Values

Query id: 6cc1ec2c-f9e4-4f5a-b547-24b196845c6a

Ok.

1 row in set. Elapsed: 0.026 sec. 

dell9510 :) insert into system.zookeeper (name) values ('x')

INSERT INTO system.zookeeper (name) FORMAT Values

Query id: eeaeb659-56ce-4272-a68e-949099c42ea9

Ok.

1 row in set. Elapsed: 0.019 sec. 

dell9510 :) insert into system.zookeeper (name, path) values ('x', '')

INSERT INTO system.zookeeper (name, path) FORMAT Values

Query id: 1fc282fa-e6c4-4160-b7d3-b7613925a603

Ok.

1 row in set. Elapsed: 0.019 sec.
```

Queries like this should throw an exception.

hanfei1991 commented:
Oops, I forgot to check the schema. Will fix it later.

On Wed, 1 Jun 2022 at 21:00, Alexander Tokmakov ***@***.***>
wrote:

> Hm, I've downloaded build from this PR, tried it locally and found some
> unexpected behavior:
>
> dell9510 :) insert into system.zookeeper (name, version, czxid) values ('x', 1, 42)
>
> INSERT INTO system.zookeeper (name, version, czxid) FORMAT Values
>
> Query id: 6cc1ec2c-f9e4-4f5a-b547-24b196845c6a
>
> Ok.
>
> 1 row in set. Elapsed: 0.026 sec.
>
> dell9510 :) insert into system.zookeeper (name) values ('x')
>
> INSERT INTO system.zookeeper (name) FORMAT Values
>
> Query id: eeaeb659-56ce-4272-a68e-949099c42ea9
>
> Ok.
>
> 1 row in set. Elapsed: 0.019 sec.
>
> dell9510 :) insert into system.zookeeper (name, path) values ('x', '')
>
> INSERT INTO system.zookeeper (name, path) FORMAT Values
>
> Query id: 1fc282fa-e6c4-4160-b7d3-b7613925a603
>
> Ok.
>
> 1 row in set. Elapsed: 0.019 sec.
>
> Queries like this should throw an exception.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ClickHouse/ClickHouse/pull/37596#issuecomment-1143579630>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE7KYQW7GN2XJ3DDOLGDJPDVM5NGDANCNFSM5XD4MIEQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

hanfei1991 committed:
try to fill in right metadata columns

hanfei1991 commented:
> Hm, I've downloaded build from this PR, tried it locally and found some unexpected behavior:
> 
> ```
> dell9510 :) insert into system.zookeeper (name, version, czxid) values ('x', 1, 42)
> 
> INSERT INTO system.zookeeper (name, version, czxid) FORMAT Values
> 
> Query id: 6cc1ec2c-f9e4-4f5a-b547-24b196845c6a
> 
> Ok.
> 
> 1 row in set. Elapsed: 0.026 sec. 
> 
> dell9510 :) insert into system.zookeeper (name) values ('x')
> 
> INSERT INTO system.zookeeper (name) FORMAT Values
> 
> Query id: eeaeb659-56ce-4272-a68e-949099c42ea9
> 
> Ok.
> 
> 1 row in set. Elapsed: 0.019 sec. 
> 
> dell9510 :) insert into system.zookeeper (name, path) values ('x', '')
> 
> INSERT INTO system.zookeeper (name, path) FORMAT Values
> 
> Query id: 1fc282fa-e6c4-4160-b7d3-b7613925a603
> 
> Ok.
> 
> 1 row in set. Elapsed: 0.019 sec.
> ```
> 
> Queries like this should throw an exception.

I marked other columns as `Materialize` to prohibit writing, and allow insert without value, because I think `insert(name, path)` means creating a path. PTAL

tavplubix commented:
Stateless tests flaky check (address, actions) - OK
Stress test - #37738

tavplubix commented on src/Storages/System/StorageSystemZooKeeper.cpp:
```diff
@@ -28,6 +32,157 @@ namespace ErrorCodes
+
+SinkToStoragePtr StorageSystemZooKeeper::write(const ASTPtr &, const StorageMetadataPtr &, ContextPtr context)
+{
+    if (!context->getConfigRef().getBool("allow_zookeeper_write", false))
```
See the discussion above
